File: app/app/src/main/java/com/memoryos/app/data/model/Item.kt
package com.memoryos.app.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.google.gson.annotations.SerializedName

@Entity(tableName = "items")
data class Item(
    @PrimaryKey
    @SerializedName("_id") val id: String = "",
    val user: String? = null,
    val url: String = "",
    val title: String = "Untitled",
    val description: String? = null,
    val domain: String? = null,
    val favicon: String? = null,
    val type: String = "page",
    val notes: String? = null,
    val content: String? = null,
    val summary: String? = null,
    val collection: String? = null,
    val duplicateOf: String? = null,
    val reviewCount: Int = 0,
    val nextReviewDate: String? = null,
    val easeFactor: Double = 2.5,
    val interval: Int = 0,
    val isTask: Boolean = false,
    val taskDone: Boolean = false,
    val favorite: Boolean = false,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    // Room can't store lists directly, so these are handled via TypeConverters
    @androidx.room.Ignore
    val tags: List<String> = emptyList()
    @androidx.room.Ignore
    val aiTags: List<String> = emptyList()
    @androidx.room.Ignore
    val keyPoints: List<String> = emptyList()
    @androidx.room.Ignore
    val highlights: List<Highlight> = emptyList()
}

data class Highlight(
    val text: String = "",
    val color: String = "#e8b931",
    val note: String? = null
)

data class ItemsResponse(
    val items: List<Item>,
    val total: Int,
    val page: Int,
    val pages: Int
)

data class ItemStats(
    val total: Int = 0,
    val favorites: Int = 0,
    val types: Map<String, Int> = emptyMap()
)

data class TagCount(
    val tag: String,
    val count: Int
)

data class CreateItemRequest(
    val url: String,
    val title: String? = null,
    val description: String? = null,
    val domain: String? = null,
    val favicon: String? = null,
    val type: String = "page",
    val tags: List<String>? = null,
    val notes: String? = null,
    val content: String? = null,
    val highlights: List<Highlight>? = null,
    val collection: String? = null,
    val isTask: Boolean = false
)

data class DuplicateWarningItem(
    @SerializedName("_id") val id: String,
    val url: String,
    val title: String,
    val domain: String?
)

data class ItemWithDuplicateWarning(
    @SerializedName("_id") val id: String,
    val url: String,
    val title: String,
    val duplicateWarning: List<DuplicateWarningItem>? = null
)

File: src/models/Item.ts
import mongoose, { Model } from "mongoose";

export interface IItem {
  _id: mongoose.Types.ObjectId;
  user: mongoose.Types.ObjectId;
  url: string;
  title: string;
  description: string;
  domain: string;
  favicon: string;
  type: "page" | "video" | "tweet" | "pdf" | "note" | "screenshot" | "transcript";
  tags: string[];
  aiTags: string[];
  notes: string;
  content: string;
  summary: string;
  keyPoints: string[];
  collection: mongoose.Types.ObjectId | null;
  duplicateOf: mongoose.Types.ObjectId | null;
  highlights: { text: string; color: string; note: string }[];
  reviewCount: number;
  nextReviewDate: Date | null;
  easeFactor: number;
  interval: number;
  isTask: boolean;
  taskDone: boolean;
  favorite: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const itemSchema = new mongoose.Schema<IItem>(
  {
    user: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },
    url: { type: String, required: true },
    title: { type: String, default: "Untitled" },
    description: { type: String, default: "" },
    domain: { type: String, default: "" },
    favicon: { type: String, default: "" },
    type: {
      type: String,
      enum: ["page", "video", "tweet", "pdf", "note", "screenshot", "transcript"],
      default: "page",
    },
    tags: [{ type: String, lowercase: true, trim: true }],
    aiTags: [{ type: String, lowercase: true, trim: true }],
    notes: { type: String, default: "" },
    content: { type: String, default: "" },
    summary: { type: String, default: "" },
    keyPoints: [{ type: String }],
    collection: { type: mongoose.Schema.Types.ObjectId, ref: "Collection", default: null },
    duplicateOf: { type: mongoose.Schema.Types.ObjectId, ref: "Item", default: null },
    highlights: [
      {
        text: { type: String, required: true },
        color: { type: String, default: "#e8b931" },
        note: { type: String, default: "" },
      },
    ],
    reviewCount: { type: Number, default: 0 },
    nextReviewDate: { type: Date, default: null },
    easeFactor: { type: Number, default: 2.5 },
    interval: { type: Number, default: 0 },
    isTask: { type: Boolean, default: false },
    taskDone: { type: Boolean, default: false },
    favorite: { type: Boolean, default: false },
  },
  { timestamps: true }
);

itemSchema.index({ user: 1, createdAt: -1 });
itemSchema.index({ user: 1, tags: 1 });
itemSchema.index({ user: 1, aiTags: 1 });
itemSchema.index({ user: 1, collection: 1 });
itemSchema.index({ user: 1, nextReviewDate: 1 });
itemSchema.index({ user: 1, isTask: 1, taskDone: 1 });
itemSchema.index({ url: 1, user: 1 });
itemSchema.index({ title: "text", description: "text", notes: "text", content: "text", summary: "text" });

const Item: Model<IItem> =
  mongoose.models.Item || mongoose.model<IItem>("Item", itemSchema);

export default Item;
